src/contract.rs

```diff
@@ -0,0 +1,851 @@
+use std::fmt;
+
+use cosmwasm_std::{
+    attr, coins, entry_point, to_binary, BankMsg, Binary, Deps, DepsMut, Env, MessageInfo,
+    Response, StdResult, Uint128,
+};
+use provwasm_std::{ProvenanceMsg, ProvenanceQuery};
+
+use crate::error::ContractError;
+use crate::msg::{ExecuteMsg, QueryMsg, Validate};
+use crate::state::{config_read, get_invoice_storage, get_invoice_storage_read, Invoice};
+
+pub const CRATE_NAME: &str = env!("CARGO_CRATE_NAME");
+pub const PACKAGE_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+// smart contract execute entrypoint
+#[entry_point]
+pub fn execute(
+    deps: DepsMut<ProvenanceQuery>,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<ProvenanceMsg>, ContractError> {
+    msg.validate()?;
+
+    match msg {
+        ExecuteMsg::AddInvoice {
+            id,
+            amount,
+            description,
+        } => add_invoice(deps, info, id, amount, description),
+        ExecuteMsg::CancelInvoice { id } => cancel_invoice(deps, info, id),
+        ExecuteMsg::PayInvoice { id } => pay_invoice(deps, info, id),
+    }
+}
+
+fn add_invoice(
+    deps: DepsMut<ProvenanceQuery>,
+    info: MessageInfo,
+    id: String,
+    amount: Uint128,
+    description: Option<String>,
+) -> Result<Response<ProvenanceMsg>, ContractError> {
+    // get state for auth and attrs
+    let state = &config_read(deps.storage).load()?;
+
+    // ensure message sender is admin
+    if info.sender != state.admin {
```

Comment:
definitely a downside of having to migrate all sc instances with pio upgrades

hoping this sc is simple enough that outside of cosmwasm upgrades it doesn't change very frequently but you bring up a good point that it requires everyone upgrade

i'm not sure how we get around that though given that this won't be centralized by figure pay since idea is that any merchant could implement this and receive pay network payments